packages/expo-cli/src/commands/build/ios/IOSBuilder.ts

```diff
@@ -119,9 +111,102 @@ class IOSBuilder extends BaseBuilder {
+  // All config validation should happen here before any build logic takes place.
+  // It's important that the errors are revealed in a thoughtful manner.
+  async checkProjectConfig(): Promise<void> {
+    // Run this first because the error messages are related
+    // to ExpoKit which is harder to change than the bundle ID.
+    await super.checkProjectConfig();
+
+    // Check the SDK version next as it's the second hardest thing to change.
+    const sdkVersion = this.manifest.sdkVersion;
+
+    await utils.checkIfSdkIsSupported(sdkVersion!, PLATFORMS.IOS);
+
+    // Validate the icon third since it's fairly easy to modify.
+    await this.validateIcon();
+
+    // Check the bundle ID and possibly prompt the user to add a new one.
+    const bundleIdentifier = this.manifest.ios?.bundleIdentifier;
+
+    if (!bundleIdentifier) {
+      // Recommend a bundle ID based on the username and project slug.
+      const username = await this.getUsernameAsync();
+      const recommendedBundleId = username ? `com.${username}.${this.manifest.slug}` : undefined;
+
+      log.newLine();
+      log(
+        log.chalk.cyan(
+          `Now we need to know your ${terminalLink(
+            'iOS bundle identifier',
+            'https://expo.fyi/bundle-identifier'
+          )}.\nYou can change this in the future if you need to.`
+        )
+      );
+      log.newLine();
+
+      // Prompt the user for the bundle ID.
+      // Even if the project is using a dynamic config we can still
+      // prompt a better error message, recommend a default value, and help the user
+      // validate their custom bundle ID upfront.
+      const bundleIdPrompt = await prompt(
+        [
+          {
+            name: 'bundleIdentifier',
+            default: recommendedBundleId,
+            // The Apple helps people know this isn't an EAS feature.
+            message: `What would you like your  bundle identifier to be?`,
+            validate: (value: string) => /^[a-zA-Z][a-zA-Z0-9\-.]+$/.test(value),
+          },
+        ],
+        {
+          nonInteractiveHelp: noBundleIdMessage,
+        }
+      );
+
+      const modification = await modifyConfigAsync(
+        this.projectDir,
+        { ios: { bundleIdentifier: bundleIdPrompt.bundleIdentifier } },
+        { skipSDKVersionRequirement: true }
+      );
+      if (modification.type === 'success') {
+        log.newLine();
+        // Success!
+        log(`Your  bundle identifier is now: ${bundleIdPrompt.bundleIdentifier}`);
+        log.newLine();
+      } else {
+        log.newLine();
+        if (modification.type === 'warn') {
+          // The project is using a dynamic config, give the user a helpful log and bail out.
+          log(log.chalk.yellow(modification.message));
+        } else {
+          log(
+            log.chalk.yellow(
+              'No Expo config was found. Please create an `app.config.json` in your project root.'
```

Comment:
```suggestion
              'No Expo config was found. Please create an `app.config.json` in your project root.'
```

i think this is kind of confusing - in our docs we generally don't refer to it as app.config.json. maybe we can provide some helper function that just says "Expo config" and terminal links to a fyi page about what that means